Implement support for mobile redirection

pierlon opened the pull request:
# Summary

* Implement support for redirecting mobile visitors to the AMP version. Fixes #4479.
  * Add `MobileRedirection` service.
  * Mobile redirection only is enabled when not in Standard mode.
  * Adds a new option on the 'AMP Settings' page to configure mobile redirection when in Transitional or Reader mode. 
  * Automatically enable AMP-to-AMP linking when mobile redirection is enabled.
  * When on mobile, add link to footer which allows a user to go to the non-AMP version (and discontinue automatic redirection), and when on the non-AMP version to go back to the AMP version (and resume automatic redirection).
  * By default use client-side redirection via JavaScript to redirect to the AMP version when a mobile user agent is detected. Use server-side redirection when `amp_mobile_client_side_redirection` is filtered `false`. Please note that this has no effect when in ~AMP Dev Mode or~ the Customizer preview, as in these cases JS is required in order to prevent the link from functioning at all; an alert is shown instead. 
  * Mobile redirection is disabled via `sessionStorage` flag during client-side redirection (thus no GDPR concerns), whereas a cookie is necessarily used when server-side redirection is being done.
  * Add default styling for the mobile version switcher link which can be disabled via `add_filter( 'amp_mobile_version_switcher_styles_used', '__return_false' )`.
  * The link text can be filtered via `amp_mobile_version_switcher_link_text`.
  * Eliminate old Exit Reader Mode link in classic template.
  * Prevent linking to AMP version when the URL has `?noamp`, in addition to whether it has `rel=noamphtml`.
  * Reuse the user agents listed in `wp_is_mobile()` to determine whether the request is a mobile user agent, but pass the user agents through an `amp_mobile_user_agents` filter to customize. Each string may be either a simple needle to search in the `User-Agent` haystack, or if it matches the pattern `^/.+/i?$` then it will be used as a regular expression match. These same user agent strings are used in both the PHP user-agent detection and the JS user-agent detection.
  * When server-side redirection is being performed, the `amp_pre_is_mobile` filter is used to override whether the user agent should be considered mobile.
  * Send `Vary: User-Agent` response header when server-side redirection is being performed.
  * Add a [`rel=alternate` link](https://developers.google.com/search/mobile-sites/mobile-seo/separate-urls#annotation-in-the-html) pointing to the AMP version to go along with the `rel=amphtml` link already on the page.
* Introduce `is_amp_available()` function to serve as subset of `is_amp_endpoint()`, consolidating and removing duplicated logic. Fixes #4619.
* Add filter to exclude links on an individual basis from AMP-to-AMP linking. Fixes #4913.
  *  Introduces `amp_to_amp_linking_element_excluded` filter which serves as the last word to override the `amp_to_amp_excluded_urls` filter.
  * An internal `amp_to_amp_linking_element_query_vars` filter is added which allows query vars to be added to the links/forms which are excluded from AMP-to-AMP linking (for example to add `noamp` query var to prevent mobile redirection).
  * Improves logic for handling AMP-to-AMP linking, including excluding `form` elements which have `rel=noamphtml`.
* Prevent automatically redirecting AMP to non-AMP for administrative users. Fixes #4626.
  * When in paired AMP, show dirty AMP page with ❌ in admin bar as opposed to redirecting to non-AMP version (since user has ability to validate and get the current validation results). This will improve navigating the site in paired browsing.
  * Show summary of the validation errors which are kept and unreviewed in the admin bar.
  * Move more validation logic from `AMP_Theme_Support` to `AMP_Validation_Manager`.
  * When redirecting from AMP to non-AMP because of kept invalid markup, add `noamp=available` which causes AMP to then not be available on the redirected page (similar to exiting the mobile version).
* Improve services infrastructure.
  * Add `amp_default_options` filter for services to inject their own service defaults for `amp-options`. Use for Plugin Suppression.
* Fix ability to preview changes to the validation error statuses.
* Allow AMP to be disabled for a request via `noamp=available` when user can validate. Also add a "View with AMP disabled" link to the admin bar for such users to easily disable AMP for a URL when a site is in Standard mode without requiring them to switch the entire site to Transitional mode or to disable AMP for a given URL. See #1294.

## Todo
- [x] Add an option to the admin screen to manage mobile redirection
    - [x] Only display option when Transitional or Reader mode is selected
- [x] Remove Customizer setting to show "Exit reader mode" in header template for Reader mode
- [x] Redirect non-AMP to AMP
- [x] Add a link to the footer to exit mobile version when on AMP page
- [x] Add a link to the footer to go to mobile version when on non-AMP page
- [x] Prevent redirection to AMP version if `?noamp=1` is present
    - [x] Prevent further redirections when navigating from the current non-AMP page
- [x] Default to JS mobile redirection, with an option to opt-in to a server-side solution instead
- [x] Automatically add `?noamp=1` to any intra-site link URLs that AMP-to-AMP linking skips
- [x] Prevent redirection to AMP page if there are kept validation errors
- [x] Ensure that `Vary: User-Agent` response header is added when redirection is enabled.
- [x] Enable AMP-to-AMP linking when mobile redirection is enabled.
- [x] Add `link[rel=alternate]` when redirection is enabled?
- [x] Determine best way to handle redirection from AMP version to non-AMP version due to validation errors with kept markup (when the current user can't validate).
- [x] Include `noamp` in removable query vars.
- [ ] Add filter to control whether or not the footer link is displayed? Currently there is a filter for excluding just the styles, but the entire link should be considered.
- [ ] Do we really need to worry about preventing AMP-to-AMP linking when noamp is in the query parameter? We could just honor the noamphtml link relation instead.
- [x] Add tests.

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

pierlon committed:
Redirect to AMP version if on mobile

pierlon committed:
Add mobile section to Customizer for non-canonical modes

pierlon committed:
Add setting for mobile redirect

pierlon committed:
Hide script if mobile redirect setting is disabled

pierlon committed:
Move "Exit mobile version" link to footer template in reader mode

pierlon committed:
Hide mobile redirect setting if standard mode is selected

pierlon committed:
Add text field to configure exit link text

pierlon committed:
Fix lint errors

westonruter commented on includes/options/class-amp-options-manager.php:
```diff
@@ -29,6 +29,7 @@ class AMP_Options_Manager {
 		Option::SUPPORTED_POST_TYPES    => [ 'post' ],
 		Option::ANALYTICS               => [],
 		Option::ALL_TEMPLATES_SUPPORTED => true,
+		Option::MOBILE_REDIRECT         => true,
```
```suggestion
		Option::MOBILE_REDIRECT         => false,
```

westonruter commented on includes/admin/class-amp-customizer.php:
```diff
@@ -13,7 +15,7 @@
  * @since 0.4
  */
-class AMP_Template_Customizer {
+class AMP_Customizer {
```
Why rename the class? Perhaps the filename should be renamed instead?

pierlon replied:
Yes. In retrospect the class name was correct. Will revert this change.

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -299,6 +299,85 @@ function amp_is_canonical() {
+		(function ( ampSlug ) {
+			var hasTouchScreen;
+
+			if ("maxTouchPoints" in navigator) {
+				hasTouchScreen = navigator.maxTouchPoints > 0;
+			} else if ("msMaxTouchPoints" in navigator) {
+				hasTouchScreen = navigator.msMaxTouchPoints > 0;
+			} else {
+				var mQ = window.matchMedia && matchMedia("(pointer:coarse)");
+				if (mQ && mQ.media === "(pointer:coarse)") {
+					hasTouchScreen = !!mQ.matches;
+				} else if ('orientation' in window) {
+					hasTouchScreen = true; // deprecated, but good fallback
+				} else {
```
I'm wary of this logic, as won't it mean laptops with touch screens will get redirected to the AMP version? If the user agent isn't used, then the minimum viewport dimension should be considered. Another option at some point would be to use `NetworkInformation.effectiveType`. Nevertheless, if a mobile visitor does land on the non-AMP version then they'll have the overhead of having to download the entire non-AMP HTML page to then get redirected to the AMP version. This being the case, it would actually be nice if a site could opt-in to doing the redirection server-side when `wp_is_mobile()` is true. The issue is that we can't rely on this by default due to variability in the network stack (as I understand).

pierlon replied:
> I'm wary of this logic, as won't it mean laptops with touch screens will get redirected to the AMP version? If the user agent isn't used, then the minimum viewport dimension should be considered.

Yes, the viewport dimensions would definitely be taken into consideration as well. I think it would be best to settle on using the max viewport width of the browser to determine whether it's a mobile device or not. Mobile screens are getting bigger by the day so we'd have to consider a sane value to compare against. 

>  This being the case, it would actually be nice if a site could opt-in to doing the redirection server-side when wp_is_mobile() is true.

Perhaps we could add a filter here before the JS script is added to let users have the option of redirecting themselves. It would be better to provide a guaranteed solution here than a variant one.

westonruter replied:
Yeah, this is tricky as it involves adding a filter to allow injecting JS code. If we provide a filter, I'd be more inclined to allow them to filter the list of user agents string tokens (or rather regular expression patterns) that cause redirection. This could then be exported to JS for use client-side but then also used to filter `wp_is_mobile` in case they opt-in to server-side redirection instead. Speaking of which, we should really provide a mechanism to do the server-side redirection if possible because redirection client-side is extremely slow, even though it is much more reliable. For sites that have administrators which are equipped technically able to verify that caching layers won't break server-side redirection, then they should be able to easily opt-in to server-side redirection via a filter. All this leads me to think that we really should limit detection to use agent sniffing, even though this has its own downsides.

westonruter commented on includes/options/class-amp-options-menu.php:
```diff
@@ -232,6 +243,52 @@ public function render_theme_support() {
+		if ( $can_customize_mobile_settings ) :
+			$mobile_customizer_url = add_query_arg(
+				[
+					'autofocus[section]' => 'amp_mobile',
```
This is tricky because in the new Reader mode when a theme is selected, this section will not exist. I wonder actually if we shouldn't just omit the setting altogether for now, but allow it to disabled via filter.

westonruter replied:
In other words, this PR could eliminate the Exit Reader Mode link, the toggle for this setting, and any toggle for showing the “Leave mobile site” link in the footer. Should this be a decision we make, as opposed to providing a option?

westonruter replied:
The "leave mobile site" link can be displayed unconditionally in the footer of the AMP version. Then in the non-AMP version the footer link can be displayed conditionally based on whether the user has mobile redirection enabled. This would make it simpler. (I realize this deviates from the requirements I laid out, but after thinking about it more I think this would be a good change.)

pierlon replied:
> This is tricky because in the new Reader mode when a theme is selected, this section will not exist. I wonder actually if we shouldn't just omit the setting altogether for now, but allow it to disabled via filter.

Disabling via a filter would make the most sense here then.

---

The new Reader mode will effectively render the current AMP Customizer panel useless, correct? If so, then I would suggest removing it now then.

pierlon replied:
> The "leave mobile site" link can be displayed unconditionally in the footer of the AMP version. Then in the non-AMP version the footer link can be displayed conditionally based on whether the user has mobile redirection enabled.

OK sounds good. We can filter showing the exit link and the text as well then.

westonruter replied:
> The new Reader mode will effectively render the current AMP Customizer panel useless, correct? If so, then I would suggest removing it now then.

This is a bit tricky. The Reader mode will effectively render the AMP Customizer panel useless _if_ a Reader theme is selected. However, it is still needed if the Classic theme is selected.

> The new Reader mode will effectively render the current AMP Customizer panel useless, correct? If so, then I would suggest removing it now then.

The exit link in the header? I propose actually that we remove it entirely in favor of only having the link in the footer. The header exit link is problematic for other reasons, e.g. #4573.

pierlon replied:
> The exit link in the header?

No I meant the one in the footer. Sorry for the brevity there.

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -299,6 +299,85 @@ function amp_is_canonical() {
+ *
+ * @return bool True if there is an AMP version, false otherwise.
+ */
+function amp_version_is_available() {
```
See #4619 and #4620

pierlon replied:
Would you consider having #4620 complete first, or port over the changes from that PR to here?

westonruter replied:
I'm happy if you want to incorporate the changes here. I've closed that PR.

pierlon committed:
Set mobile redirect option to be false

pierlon committed:
Rename AMP_Customizer class to AMP_Template_Customizer

pierlon committed:
Incorporate changes from #4620

pierlon committed:
Remove mobile section in Customizer

pierlon committed:
Add mobile version switcher markup if mobile redirection enabled

pierlon committed:
Prevent AMP to AMP linking if href contains `noamp` query parameter

pierlon commented on includes/class-amp-theme-support.php:
```diff
@@ -397,12 +404,52 @@ public static function finish_init() {
 				self::redirect_non_amp_url( current_user_can( 'manage_options' ) ? 302 : 301 );
 			}
 
+			if ( AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT ) && wp_is_mobile() && is_amp_available() ) {
```
I'm using `wp_is_mobile()` as a placeholder for now. We could even apply our own filter to that to add more stringent checks.

westonruter replied:
Where did the JS-based redirection go? That needs to remain in-place as the default redirection method. We'll only provide server-side redirection as an opt-in for sites that can safely implement it. There is another implication if going with the server-side route: namely, instead of using `sessionStorage` well need to use a cookie to capture whether the user has opted-out of the mobile site. This same cookie can be used both client-side and, if they enable server-side redirection, server-side as well.

pierlon replied:
> Where did the JS-based redirection go?

I misunderstood your reply in https://github.com/ampproject/amp-wp/pull/4796#discussion_r432678563. In https://github.com/ampproject/amp-wp/pull/4796/commits/0da5f23bf551c900cbdcca8f648d48576e89b232, I've re-added the JS solution, with the ability to opt-in to server-side redirection via a filter.

pierlon committed:
Add `noamp` query parameter to links opted out from AMP to AMP linking if mobile redirection is enabled

pierlon committed:
Add mobile redirect manager class to handle mobile redirection

pierlon committed:
Fix tests

pierlon committed:
Test disabling mobile redirect for excluded URLs and links with `noamphtml` relationship

pierlon committed:
Merge branch 'develop' into add/4479-mobile-redirection

pierlon commented:
This is available for review, albeit some tests being missing.

westonruter commented on includes/admin/class-amp-template-customizer.php:
```diff
@@ -59,6 +59,16 @@ public static function init( $wp_customize ) {
+		 * @param WP_Customize_Manager $manager Manager.
+		 */
+		do_action( 'amp_customizer_enqueue_scripts', $self->wp_customize );
+
```
Can this go back to where it was before?

pierlon replied:
Yep will revert this.

westonruter commented on includes/admin/class-amp-template-customizer.php:
```diff
@@ -133,98 +143,78 @@ public function register_settings() {
 	 * @since 0.6
 	 */
 	public function add_customizer_scripts() {
-		if ( ! amp_is_canonical() ) {
```
Is this diff chunk still relevant?

pierlon replied:
Nope, will revert.

westonruter commented on includes/admin/functions.php:
```diff
@@ -96,7 +96,7 @@ function amp_admin_get_preview_permalink() {
 function amp_add_customizer_link() {
 	/** This filter is documented in includes/settings/class-amp-customizer-design-settings.php */
-	if ( ! apply_filters( 'amp_customizer_is_enabled', true ) || current_theme_supports( AMP_Theme_Support::SLUG ) ) {
+	if ( ! apply_filters( 'amp_customizer_is_enabled', true ) || amp_is_canonical() ) {
```
I think this change can be reverted because there should not be an AMP panel in the Customizer if the theme has AMP support. In other words, `amp` theme support means that it is Standard or Transitional mode. Only in classic Reader mode is there an AMP panel.

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -301,6 +301,56 @@ function amp_add_frontend_actions() {
+			__FUNCTION__,
+			esc_html(
+				sprintf(
+						/* translators: %1$s: function name, %2$s: the current action, %3$s: the wp action, %4$s: the WP_Query class, %5$s: the amp_skip_post() function */
```
```suggestion
					/* translators: %1$s: function name, %2$s: the current action, %3$s: the wp action, %4$s: the WP_Query class, %5$s: the amp_skip_post() function */
```

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -495,37 +545,37 @@ function amp_remove_endpoint( $url ) {
 	 */
-	if ( false === apply_filters( 'amp_frontend_show_canonical', true ) ) {
-		return;
+	if ( has_filter( 'amp_frontend_show_canonical' ) ) {
+		_deprecated_hook( 'amp_frontend_show_canonical', '1.6', 'amp_frontend_show_amphtml' );
```
I don't think this is correct usage. You should rather just use `apply_filters_deprecated()` below.

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -495,37 +545,37 @@ function amp_remove_endpoint( $url ) {
-	} elseif ( $wp_query instanceof WP_Query && ( $wp_query->is_singular() || $wp_query->is_posts_page ) && post_supports_amp( get_post( get_queried_object_id() ) ) ) {
-		$amp_url = amp_get_permalink( get_queried_object_id() );
+	if (
+		false === apply_filters( 'amp_frontend_show_amphtml', true ) ||
+		false === apply_filters( 'amp_frontend_show_canonical', true )
```
These are missing the PhpDoc.

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -495,37 +545,37 @@ function amp_remove_endpoint( $url ) {
-	} elseif ( $wp_query instanceof WP_Query && ( $wp_query->is_singular() || $wp_query->is_posts_page ) && post_supports_amp( get_post( get_queried_object_id() ) ) ) {
-		$amp_url = amp_get_permalink( get_queried_object_id() );
+	if (
+		false === apply_filters( 'amp_frontend_show_amphtml', true ) ||
```
I wonder if we should even consider this in this PR. I don't think that the filter is even the right way to suppress the `amphtml` link. If anything, the better way would seem to be filtering the output of `is_amp_available()`. This would then act at a lower level, similar to `amp_skip_post`. It would actually get to implement #2817.

pierlon replied:
OK that sounds like a reasonable plan. I would recommend deprecating the `amp_frontend_show_canonical` filter, and filtering the output of `is_amp_available()` as suggested.

pierlon replied:
I've reverted the changes here via https://github.com/ampproject/amp-wp/pull/4796/commits/ccddfb26078bb2d24d4b0d6ce20ef05ecf13ef00.

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -301,6 +301,56 @@ function amp_add_frontend_actions() {
+	if ( current_theme_supports( AMP_Theme_Support::SLUG ) ) {
+		// Abort if in Transitional mode and AMP is not available for the URL.
+		$availability = AMP_Theme_Support::get_template_availability( $wp_query );
+		return $availability['supported'];
```
This isn't actually the final word… 

Consider the case in Transitional mode where there is invalid markup that is kept. In that case, AMP should actually be blocked entirely, as seen in `amp_add_amphtml_link()`:

https://github.com/ampproject/amp-wp/blob/722f8c9612b2e831217ed9e72ba50caa421f3553/includes/amp-helper-functions.php#L531-L552

westonruter replied:
In the same way, when someone does end up on a paired AMP page that has kept invalid markup, a redirect is performed to the non-AMP version:

https://github.com/ampproject/amp-wp/blob/722f8c9612b2e831217ed9e72ba50caa421f3553/includes/class-amp-theme-support.php#L2114-L2125

This URL will need to have the `noamp` query param added to prevent an infinite loop.

This would be addressing this todo here: https://gist.github.com/westonruter/5f4f27ed3bdd676e3f08524b9bab8bb4#file-amp-mobile-js-redirection-php-L103

pierlon replied:
Addressed in https://github.com/ampproject/amp-wp/pull/4796/commits/85ab507c2149ebdb93f5e4ee8b48472888128c2b.

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,183 @@
+			'Opera Mobi',
+		];
+
+		return apply_filters( 'amp_mobile_user_agents', $default_user_agents );
```
Need to add phpdoc for the filter.

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,183 @@
+	 * @return bool True if current request if from a mobile device, otherwise false.
+	 */
+	public static function is_mobile() {
+		$pre_is_mobile = apply_filters( 'amp_pre_is_mobile', false );
+
+		if ( true === $pre_is_mobile ) {
+			return (bool) $pre_is_mobile;
+		}
```
Should this be here? There'd be no way to replicate this in JS.

pierlon replied:
The reasoning here is that the user may have already determined that the current request is from a mobile device, maybe from a reverse proxy or cookie, for example. If that's the case then the server-side logic below would become redundant.

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,183 @@
+			}
+		}
+
+		return apply_filters( 'amp_is_mobile', $is_mobile );
```
Does this need to be filtered? There'd be no way to replicate this in JS.

pierlon replied:
This filter can be removed.

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,183 @@
+			} )(
+					<?php echo wp_json_encode( amp_get_slug() ); ?>,
+					<?php echo wp_json_encode( self::DISABLED_COOKIE_NAME ); ?>,
+					<?php echo '[' . implode( ',', array_map( 'wp_json_encode', self::get_user_agents() ) ) . ']'; ?>
```
```suggestion
					<?php echo wp_json_encode( self::get_user_agents() ); ?>
```

pierlon committed:
Fix lint errors

pierlon committed:
Revert changes; apply suggestions from Weston

pierlon committed:
Add docs for filters in MobileRedirectManager class

pierlon committed:
Re-enable mobile redirection for the session if the user opts to view the mobile version

pierlon committed:
Revert changes to `amp_frontend_show_canonical` filter

pierlon committed:
Prevent a infinite loop request if there are validation errors

pierlon committed:
Fix lint errors

pierlon committed:
Check for validation errors before checking if AMP is available

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,225 @@
+		</style>
+		<div id="site-version-switcher" <?php echo esc_attr( $hidden_attr ); ?>>
+			<a
+				id="version-switch-link"
+				rel="<?php echo esc_attr( $rel ); ?>"
+				href="<?php echo esc_url( $url ); ?>"
```
As discussed, if `$is_amp` then this needs to have an `onclick` attribute which clears the cookie so that mobile redirection will be resumed.

pierlon replied:
Done in 831726e399559c1b3fd17d023f23d95cdc2f8d50.

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -301,6 +301,68 @@ function amp_add_frontend_actions() {
+		return false;
+	}
+
+	return true;
```
As discussed we will probably need to filter this to allow plugins to force AMP to be enabled or disabled ona  per-URL basis. See #2817.

Should this filter allow overriding all the above `return false` statements as well so to force a response that would have been not AMP to be instead be forced to be AMP? This would end up being a way to implement #4626.

westonruter replied:
No need to do that now. We can revisit this later.

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -512,43 +571,40 @@ function amp_add_amphtml_link() {
-	}
-
 	// Check to see if there are known unaccepted validation errors for this URL.
 	if ( current_theme_supports( AMP_Theme_Support::SLUG ) ) {
```
Eventually this logic here should probably be moved inside of `is_amp_available()`, but we should check the implications of doing so first.

westonruter replied:
To examine later.

westonruter commented on tests/php/test-amp-helper-functions.php:
```diff
@@ -364,9 +364,9 @@ public function test_amp_add_amphtml_link_reader_mode( $data_provider ) {
 		// Make sure adding the filter hides the amphtml link.
-		add_filter( 'amp_frontend_show_canonical', '__return_false' );
+		add_filter( 'amp_frontend_show_amphtml', '__return_false' );
 		$this->assertEmpty( $get_amp_html_link() );
-		remove_filter( 'amp_frontend_show_canonical', '__return_false' );
+		remove_filter( 'amp_frontend_show_amphtml', '__return_false' );
```
These can be reverted now, correct?

westonruter commented on tests/php/test-amp-helper-functions.php:
```diff
@@ -467,9 +467,9 @@ public function test_amp_add_amphtml_link_transitional_mode( $data_provider ) {
 		// Make sure adding the filter hides the amphtml link.
-		add_filter( 'amp_frontend_show_canonical', '__return_false' );
+		add_filter( 'amp_frontend_show_amphtml', '__return_false' );
 		$this->assertEmpty( $get_amp_html_link() );
-		remove_filter( 'amp_frontend_show_canonical', '__return_false' );
+		remove_filter( 'amp_frontend_show_amphtml', '__return_false' );
```
These can be reverted now, right?

pierlon replied:
Yes, reverted them in 4412fabcf5773ea5f09c91842ab7bd28e9fedfe3.

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,225 @@
+						const versionSwitchLink = document.getElementById( 'version-switch-link' );
+						if ( versionSwitchLink ) {
+							versionSwitchLink.addEventListener( 'click', () => {
+								document.cookie = `${disabledCookieName}=;path=/;samesite=strict${ 'https:' === location.protocol ? ';secure' : '' }`;
```
The JS used here needs to be IE11-compatible, so unfortunately no backticks or arrow functions.

pierlon replied:
Done in 709ab6b687fde4f5c796309530b4bd78394b0c57.

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -397,12 +398,54 @@ public static function finish_init() {
+						[
+							'expires'  => 0, // Cookie will only last for the current browser session.
+							'path'     => '/',
+							'secure'   => isset( $_SERVER['HTTPS'] ),
```
```suggestion
							'secure'   => is_ssl(),
```

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -397,12 +398,54 @@ public static function finish_init() {
+						'1',
+						[
+							'expires'  => 0, // Cookie will only last for the current browser session.
+							'path'     => '/',
```
To account for multisite and subdirectory installs, this needs to use the home URL:

```suggestion
							'path'     => wp_parse_url( home_url( '/' ), PHP_URL_PATH ),
```

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -397,12 +398,54 @@ public static function finish_init() {
+				// Redirect if `noamp` query parameter is not present and JS redirection is disabled.
+				if ( ! $mobile_redirection_disabled && ! MobileRedirectManager::should_redirect_via_js() ) {
+					$amp_url = add_query_arg( amp_get_slug(), '1', amp_get_current_url() );
+					wp_safe_redirect( $amp_url, current_user_can( 'manage_options' ) ? 302 : 301 );
```
I think 302 is sufficient here. By virtue of the cookie being session-based, this entails a temporary redirect. The only reason for doing 301 redirect was for the sake of anyone concerned about SEO. The 302 redirect was used for people who were admins since they can change the mode. But just only doing 302 should be all that is needed here:

```suggestion
					wp_safe_redirect( $amp_url, 302 );
```

Speaking of which, I bet the link to leave the mobile site should have a `rel=nofollow`.

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -397,12 +398,54 @@ public static function finish_init() {
 		self::ensure_proper_amp_location();
 
+		if ( ! amp_is_canonical() && MobileRedirectManager::is_enabled() && is_amp_available() ) {
+			// @todo The `amp_post_template_footer` should only be used for the reader mode Classic theme.
```
Why the todo?

pierlon replied:
If I understand correctly, the `amp_post_template_footer` action is only applied in the reader mode Classic theme. If that is the case, then we will only need to add to this action if Classic is the selected reader mode theme, otherwise we would use the `wp_footer` action.

westonruter replied:
Right, but isn't that todo already implemented in the ternary below? In any case, it wouldn't hurt to add the `wp_footer` and `amp_post_template_footer` both since the two actions aren't triggered on the same template.

pierlon replied:
Yes, but given the upcoming enhancements to the reader mode, I suppose the conditional below would no longer be correct, given that there will be a variety of themes to choose from for that template mode.

westonruter replied:
Good point. In that case, let's just do two `add_action()` calls here, one for `wp_footer` and another for `amp_post_template_footer`. Only one of them will end up getting used.

pierlon replied:
Sounds good :+1: .

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -2116,6 +2179,11 @@ static function( Optimizer\Error $error ) {
 				if ( AMP_Validation_Manager::has_cap() ) {
 					$non_amp_url = add_query_arg( AMP_Validation_Manager::VALIDATION_ERRORS_QUERY_VAR, $blocking_error_count, $non_amp_url );
+
+					if ( MobileRedirectManager::is_mobile_request() ) {
+						// Disable mobile redirection to prevent an infinite loop.
+						$non_amp_url = add_query_arg( MobileRedirectManager::NO_AMP_QUERY_VAR, '1', $non_amp_url );
+					}
```
I think this is relevant whether it is a mobile request or not:

```suggestion
					// Disable mobile redirection to prevent an infinite loop.
					$non_amp_url = add_query_arg( MobileRedirectManager::NO_AMP_QUERY_VAR, '1', $non_amp_url );
```

pierlon replied:
The query var would have no effect since it is not a mobile request, no?

westonruter replied:
Yes, but only safely so if `should_redirect_via_js` is `false`. It would be safer to add the `noamp` query var always to ensure that JS-based redirection will also not end up doing a redirection.

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,225 @@
+	 *              Otherwise, returns true if mobile redirection option is enabled and current request is from a mobile device.
+	 */
+	public static function is_enabled() {
+		$option_enabled = AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT );
+
+		return self::should_redirect_via_js() ? $option_enabled : $option_enabled && self::is_mobile_request();
```
```suggestion
		if ( ! AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT ) ) {
			return false;
		}
		return self::should_redirect_via_js() || self::is_mobile_request();
```

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,225 @@
+			return (bool) $pre_is_mobile;
+		}
+
+		$current_user_agent = $_SERVER['HTTP_USER_AGENT'];
```
Because WordPress…

```suggestion
		$current_user_agent = wp_unslash( $_SERVER['HTTP_USER_AGENT'] );
```

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,225 @@
+		$user_agents = self::get_user_agents();
+
+		foreach ( $user_agents as $user_agent ) {
+			if ( false !== strpos( $current_user_agent, $user_agent ) ) {
```
We may want to consider `$user_agent` being a regular expression pattern rather than just a substring match.

pierlon replied:
Done in a7a72f83bd42e24824f3f3863e730a06ee9e4779.

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,225 @@
+
+		foreach ( $user_agents as $user_agent ) {
+			if ( false !== strpos( $current_user_agent, $user_agent ) ) {
+				$is_mobile = true;
+				break;
```
```suggestion
				return true;
```

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,225 @@
+			}
+		}
+
+		return $is_mobile;
```
```suggestion
		return false;
```

westonruter commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,225 @@
+	 * @param string $text   Text for the anchor element.
+	 */
+	public static function add_mobile_version_switcher_markup( $is_amp, $url, $text ) {
+		$rel         = $is_amp ? 'amphtml' : 'noamphtml';
```
To prevent mobile crawlers from getting stuck on the non-AMP version:

```suggestion
		$rel         = $is_amp ? 'amphtml' : 'nofollow noamphtml';
```

westonruter commented on src/Option.php:
```diff
@@ -41,6 +41,15 @@ interface Option {
+	 *
+	 * Default value: true
+	 *
+	 *@var string
```
```suggestion
	 * @var string
```

westonruter commented on src/Option.php:
```diff
@@ -41,6 +41,15 @@ interface Option {
 	const DISABLE_CSS_TRANSIENT_CACHING = 'amp_css_transient_monitor_disable_caching';
 
+	/**
+	 * Redirect mobile visitors to the AMP version of a page when the site is in Transitional or Reader mode.
+	 *
+	 * Default value: true
```
I think this may be out of date.

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -397,12 +398,54 @@ public static function finish_init() {
 			}
 
+			if ( MobileRedirectManager::is_enabled() && is_amp_available() ) {
+				$mobile_redirection_disabled = isset( $_GET[ MobileRedirectManager::NO_AMP_QUERY_VAR ] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
I suggest putting this in a method of `MobileRedirectManager`, called something like `is_request_redirection_disabled()`

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -397,12 +398,54 @@ public static function finish_init() {
 				self::redirect_non_amp_url( current_user_can( 'manage_options' ) ? 302 : 301 );
 			}
 
+			if ( MobileRedirectManager::is_enabled() && is_amp_available() ) {
```
Maybe more of the logic in this block should move to method(s) in `MobileRedirectManager`?

pierlon replied:
Done in https://github.com/ampproject/amp-wp/pull/4796/commits/5cb756ed4759dac86d1b77441d3e0e741138cac2.

pierlon committed:
Fix issues with redirection via PHP

pierlon committed:
Make redirect JS script IE11 compatible

pierlon committed:
Use regex to identify mobile user agent

pierlon committed:
Revert changes to `Test_AMP_Helper_Functions` class

pierlon committed:
Simplify args for `setcookie`

pierlon committed:
Apply suggestions from Weston

pierlon committed:
Move logic related redirection to `MobileRedirectManager` class

pierlon committed:
Fix test

pierlon committed:
Remove arrow function and string interpolation

pierlon committed:
Remove conditional determining which action to use to output the markup in the footer

pierlon committed:
Send 'Vary' header

schlessera commented on includes/admin/class-amp-template-customizer.php:
```diff
@@ -134,11 +134,16 @@ public function register_settings() {
 				amp_get_asset_url( 'js/amp-customize-controls.js' ),
-				[ 'jquery', 'customize-controls' ],
-				AMP__VERSION,
+				array_merge( $dependencies, [ 'jquery', 'customize-controls' ] ),
```
I wonder whether we couldn't set up webpack so that all dependencies for JS code are properly declared within the JS source, instead of distributing the dependency knowledge across both JS and PHP. 🤔

johnwatkins0 replied:
Pretty sure this would work:

1. In webpack.config.js
```JS
const customizer = {
	...sharedConfig,
	entry: {
		'amp-customize-controls': './assets/src/customizer/amp-customize-controls.js',
		'amp-customize-preview': './assets/src/customizer/amp-customize-preview.js',
		'amp-customizer-design-preview': './assets/src/customizer/amp-customizer-design-preview.js',
	},

	plugins: [
		...sharedConfig.plugins,
		new WebpackBar( {
			name: 'Customizer',
			color: '#f27136',
		} ),
		new DependencyExtractionWebpackPlugin( {
			requestToExternal: ( request ) => {
				if ( 'customize-controls' === request ) {
					return request;
				}
				return defaultRequestToExternal( request );
			},
		} ),
	],
};
```

2. In amp-customize-controls.js

```
import jquery from 'jquery';
import customize from 'customize-controls';

// ...

}( customize, jquery ) ); // last line

```

3. Update this line to remove `[ 'jquery', 'customize-controls' ]` -- doesn't the dependency extraction Webpack plugin handle jQuery?

Then a no-unresolved ESlint exception for `customize-controls`.

schlessera commented on includes/amp-helper-functions.php:
```diff
@@ -301,6 +301,66 @@ function amp_add_frontend_actions() {
+ * @todo This should go into the AmpProject\AmpWP namespace?
+ * @since 1.6
+ *
+ * @return bool True if there is an AMP version, false otherwise.
```
This made me first read as "true if there is an AMP version _number_", instead an "AMP version of the page".

Also, explaining the concepts of true and false and their exclusivity on a boolean value seems moot.

```suggestion
 * @return bool Whether there is an AMP version for the provided URL.
```

westonruter replied:
Rather for the “current request” as opposed to the “provided URL”:

```suggestion
 * @return bool Whether AMP is available for the current request.
```

schlessera commented on includes/settings/class-amp-customizer-design-settings.php:
```diff
@@ -186,30 +171,18 @@ public static function register_customizer_ui( $wp_customize ) {
 			]
 		);
 
-		// Header.
```
```php
// Header
'amp-wp-footer'
```

😆

schlessera commented on tests/php/test-class-amp-link-sanitizer.php:
```diff
@@ -238,6 +239,50 @@ public function test_amp_to_amp_meta_tag( $sanitizer_args, $expected_meta ) {
+		add_filter( 'pre_option_' . AMP_Options_Manager::OPTION_NAME, $enable_mobile_redirect );
+
+		$link = home_url( '/' );
+		$dom  = AMP_DOM_Utils::get_dom_from_content( "<a id='link' href='$link'>Foo</a>" );
```
I suggest always using the curly brace nation for variable substitution as it is both clearer to read and avoids breaking in some edge cases:
```suggestion
		$dom  = AMP_DOM_Utils::get_dom_from_content( "<a id='link' href='{$link}'>Foo</a>" );
```

schlessera commented on tests/php/test-class-amp-link-sanitizer.php:
```diff
@@ -238,6 +239,50 @@ public function test_amp_to_amp_meta_tag( $sanitizer_args, $expected_meta ) {
+		add_filter( 'pre_option_' . AMP_Options_Manager::OPTION_NAME, $enable_mobile_redirect );
+
+		$link = home_url( '/' );
+		$dom  = AMP_DOM_Utils::get_dom_from_content( "<a id='link' href='$link' rel='noamphtml'>Foo</a>" );
```
```suggestion
		$dom  = AMP_DOM_Utils::get_dom_from_content( "<a id='link' href='{$link}' rel='noamphtml'>Foo</a>" );
```

schlessera commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,305 @@
+	 * @return bool If JS redirection is disabled, only the status of the mobile redirection option is returned.
+	 *              Otherwise, returns true if mobile redirection option is enabled and current request is from a mobile device.
+	 */
+	public static function is_enabled() {
```
Was there a technical necessity for making this entire class static? If so, what was it?

A purely static class is an anti-pattern, as it goes against OOP principles and basically behaves like a global variable with attached methods.

Also, it makes testing unnecessarily difficult. It would be preferable to have a regular object that is injected into `AMP_Theme_Support` instead, as you can then mock the object to inject and control it within the tests.

schlessera commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,305 @@
+		 *
+		 * @since 1.6
+		 *
+		 * @param null $is_mobile Whether the current request is from a mobile device.
```
```suggestion
		 * @param bool|null $is_mobile Whether the current request is from a mobile device.
```

pierlon replied:
In this case wouldn't the parameter always be null?

westonruter replied:
In the case of filter docs, I think the phpdoc needs to capture what the filtered value _could_ be, not just what it initially is. Compare with the `pre_option_{$option}` filter. It is documented as:

```
	 * @param bool|mixed $pre_option The value to return instead of the option value. This differs from
	 *                               `$default`, which is used as the fallback value in the event the option
	 *                               doesn't exist elsewhere in get_option(). Default false (to skip past the
	 *                               short-circuit).
```

pierlon replied:
OK, thanks for the clarification :+1: .

schlessera commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,305 @@
+		$pre_is_mobile = apply_filters( 'amp_pre_is_mobile', null );
+
+		if ( null !== $pre_is_mobile ) {
+			return (bool) $pre_is_mobile;
```
Casting to `(bool)` can easily cloak misuse and fail without giving a clue as to why. Maybe use `doing_it_wrong()` instead if a non-boolean is passed in?

schlessera commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,305 @@
+		 *
+		 * @param bool $should_redirect_via_js Whether JS redirection should be used.
+		 */
+		return (bool) apply_filters( 'amp_redirect_via_js', true );
```
This could also be a candidate for `doing_it_rong()` if a non-boolean is passed in.

schlessera commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,305 @@
+		 *
+		 * @param string[] $user_agents List of user agents.
+		 */
+		return apply_filters( 'amp_mobile_user_agents', $default_user_agents );
```
The method claims it will only every return an array, so it should enforce that, filters are not to be trusted.
```suggestion
		return (array) apply_filters( 'amp_mobile_user_agents', $default_user_agents );
```

schlessera commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,305 @@
+	 * @return bool True if disabled, false otherwise.
+	 */
+	public static function redirection_disabled_for_request() {
+		return ( isset( $_GET[ self::NO_AMP_QUERY_VAR ] ) && '1' === $_GET[ self::NO_AMP_QUERY_VAR ] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
`FILTER_VALIDATE_BOOLEAN` should be used here, it works with `'1'`, `'true'`, `'on'` and  `'yes'`.
```suggestion
		return filter_input( INPUT_GET, self::NO_AMP_QUERY_VAR, FILTER_VALIDATE_BOOLEAN );
```

westonruter replied:
We haven't used the `filter_input()` function in the codebase yet, in part because WordPress core does not use it since it is not always available. So if we use it, we'll have to include `filter` among the required extensions that we check for in `amp.php`.

Also, can the value be programmatically manipulated for unit testing?

schlessera commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,305 @@
+	 * @return bool True if disabled, false otherwise.
+	 */
+	public static function redirection_disabled_for_session() {
+		return ( isset( $_COOKIE[ self::DISABLED_COOKIE_NAME ] ) && '1' === $_COOKIE[ self::DISABLED_COOKIE_NAME ] ) || self::$disabled_cookie_is_set;
```
`FILTER_VALIDATE_BOOLEAN` should be used here, it works with `'1'`, `'true'`, `'on'` and  `'yes'`.

Also, the fastest check should always go first if the order doesn't matter.
```suggestion
		return self::$disabled_cookie_is_set || filter_input( INPUT_GET, self::DISABLED_COOKIE_NAME, FILTER_VALIDATE_BOOLEAN );
```

westonruter replied:
See above.

Also, as opposed to using `filter_input` we could use `rest_sanitize_boolean()`

schlessera commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,305 @@
+	public static function add_mobile_redirect_script() {
+		?>
+		<script>
+			(function ( ampSlug, disabledCookieName, userAgents ) {
```
```suggestion
			( function ( ampSlug, disabledCookieName, userAgents ) {
```

schlessera commented on src/MobileRedirectManager.php:
```diff
@@ -0,0 +1,305 @@
+						// Escape each user agent string before forming the regex expression.
+						.map( function ( userAgent ) {
+							// See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Escaping.
+							return userAgent.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
```
```suggestion
							return userAgent.replace( /[.*+\-?^${}()|[\]\\]/g, '\\$&' ); // $& means the whole matched string
```

pierlon committed:
Fix formatting issues

pierlon committed:
Remove static methods from MobileRedirectManager class

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/4479-mobile-redirection

* 'develop' of github.com:ampproject/amp-wp: (459 commits)
  Update dependency autoprefixer to v9.8.2
  Update dependency eslint-plugin-jest to v23.14.0
  Store true/false strings in usermeta for amp_dev_tools_enabled
  Opt out of using Delayed for DevToolsUserAccess
  Update Jest snapshot
  Update jsdoc for UserContextProvider; improve tech question description
  Supply DEFAULT_BASE_BRANCH in .travis.yml to fix wp-dev-lib issue
  Further fix PluginSuppressionTest
  Fix test logic for updating suppressed plugins
  Use simple comparison for version change since false causes PHP 7.4 error
  Update dependency autoprefixer to v9.8.1
  Update amphtml to v2006112352000
  Update dependency @babel/cli to v7.10.3
  Eliminate long-deprecated AMP_Blacklist_Sanitizer
  Sync amp-toolbox-test-suite
  Remove obsolete scripts to create-gutenberg-test-pos
  Eliminate removable references to ~master~ branch
  Improve variable name and comment
  Update dependency eslint to v7.3.0
  Centralize option sanitization logic into Service
  ...

westonruter committed:
Fix phpcs issues

westonruter committed:
Switch MobileRedirectionManager be implemented as a Service

westonruter committed:
Fix malformed phpdoc var tag

westonruter committed:
Remove redundant check for blocking validation errors in amp_add_amphtml_link()

This logic is now contained within is_amp_available()

westonruter committed:
Move logic from is_amp_endpoint() to is_amp_available()

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/4479-mobile-redirection

* 'develop' of github.com:ampproject/amp-wp:
  Update dependency @wordpress/e2e-test-utils to v4.9.0
  Update dependency @wordpress/components to v9.8.0
  Update dependency webpack-cli to v3.3.12
  Update dependency @wordpress/edit-post to v3.19.0
  Update dependency @babel/core to v7.10.3
  Update dependency @wordpress/block-editor to v4.1.0
  Update dependency @wordpress/eslint-plugin to v7
  Remove extraneous parenthesis
  Use quicker PHP version comparison
  Remove overridden before_script key
  Fix getting static property value
  Fix lint & phpstan errors
  Fix some errors in PHP 8
  Add mysql service
  Update dist to xenial
  Remove unmatched error pattern
  Bump some jobs to use PHP 7.4
  Update dependency eslint-plugin-jest to v23.16.0
  Add job to test on PHP 8.0
  Update dependency terser-webpack-plugin to v3

westonruter committed:
Rename MobileRedirectManager to MobileRedirection

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -431,6 +468,30 @@ static function() {
+	 * Output the markup that allows the user to switch to the non-AMP version of the page.
+	 */
+	public static function add_non_amp_mobile_version_switcher() {
+		$url = add_query_arg( MobileRedirection::NO_AMP_QUERY_VAR, '1', self::get_current_canonical_url() );
```
No need to use `get_current_canonical_url`

westonruter committed:
Move logic from amp-theme-support to mobile redirection service and validation manager

westonruter committed:
Fix ability to preview validation error status changes

westonruter committed:
Fix faulty logic to remove query param from location

westonruter committed:
Update tests so noamp=1 is always added to the unavailable AMP redirection URL

westonruter committed:
Update tests for previewing status overrides in a389209cd

westonruter committed:
Move mobile redirection JS into separate Webpack-processed file

westonruter committed:
Improve passing globals to mobile-redirection script

westonruter commented on .eslintrc:
```diff
@@ -104,6 +104,16 @@
+      "globals": {
+        "AMP_MOBILE_REDIRECTION": false,
+        "location": false,
+        "navigator": false
```
There's probably a better way to add these globals the allowlist.

westonruter commented on .eslintrc:
```diff
@@ -104,6 +104,16 @@
+        "assets/src/mobile-redirection.js"
+      ],
+      "globals": {
+        "AMP_MOBILE_REDIRECTION": false,
```
This `AMP_MOBILE_REDIRECTION` “global” is replaced with JSON data when the script contents are output. Perhaps there is a more elegant way.

westonruter committed:
Add explanation for where AMP_MOBILE_REDIRECTION comes from

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/4479-mobile-redirection

* 'develop' of github.com:ampproject/amp-wp:
  Register services right away

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/4479-mobile-redirection

* 'develop' of github.com:ampproject/amp-wp:
  Bump stable tag to 1.5.4
  Update dependency uuid to v8.2.0
  Update dependency eslint-plugin-jsdoc to v28.5.1
  Uncomment config exports
  Refactor Webpack config to remove duplicated instances of plugins and module rules
  Update schema for CopyWebpackPlugin
  Add and fix missing docs where necessary
  Update dependency babel-jest to v26.1.0
  Add indentation where necessary
  Individually check for theme support features
  Remove duplicated calls to `parent::setUp()`
  Temporarily fix issue with GB 8.3.0
  Fix lint errors
  Skip block pre-rendering for tests that are not block related
  Ignore lint error
  Fix failing tests when Gutenberg >= 8.1.0 is installed
  Download the appropriate latest version of Gutenberg

westonruter committed:
Introduce amp_default_options filter for Services to provide their default options

westonruter committed:
Move mobile redirect option default and sanitization to service

westonruter committed:
Move settings field logic to MobileRedirection service

westonruter commented on includes/options/class-amp-options-manager.php:
```diff
@@ -97,6 +96,14 @@ public static function get_options() {
+		 * @internal
+		 * @param array $defaults Default options.
+		 */
+		$defaults = apply_filters( 'amp_default_options', $defaults );
```
@schlessera FYI

westonruter committed:
Update tests for PluginSuppression

westonruter commented on assets/src/mobile-redirection.js:
```diff
@@ -0,0 +1,44 @@
+		// Escape each user agent string before forming the regex expression.
+		.map( function( userAgent ) {
+			// See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Escaping.
+			return userAgent.replace( /[.*+\-?^${}()|[\]\\]/g, '\\$&' ); // $& means the whole matched string
```
If we're escaping the pattern, then we might as well just loop over all of the `userAgents` and see if `navigator.userAgent.includes( userAgent )`. No need for regex in this case.

westonruter replied:
Done in ac3ad30ef04db05d7a352252c0b93b6365f55df1

westonruter commented on src/MobileRedirection.php:
```diff
@@ -0,0 +1,431 @@
+			)
+		);
+
+		if ( preg_match( "/$user_agents_regex/", $current_user_agent ) ) {
```
As noted in the JS file, if `$this->get_user_agents()` is a list of strings and not a list of regexes, then we might as well just use `strpos()` rather than `preg_match()`.

westonruter replied:
Done in ac3ad30ef04db05d7a352252c0b93b6365f55df1

westonruter committed:
Introduce amp_to_amp_linking_element_excluded filter to decouple MobileRedirection from A2A linking

Fixes #4913

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -2092,42 +2078,23 @@ static function( Optimizer\Error $error ) {
+			 * @internal
+			 * @param string $non_amp_url Non-AMP URL.
+			 */
+			$non_amp_url = apply_filters( 'amp_unavailable_redirect_url', $non_amp_url );
```
I want to see if we can not need this.

westonruter committed:
Eliminate unused constants REL_VALUE_AMP and REL_VALUE_NON_AMP_TO_AMP

westonruter committed:
Improve filter name: amp_to_amp_linking_element_query_vars

westonruter committed:
Discontinue redirecting admins to non-AMP version when kept invalid markup

For non-admins, redirect to non-AMP with a special value for the noamp query var that prevents AMP from even being made available (so there is no link to try to go back to the AMP version).

Show the status of the validation errors in the admin bar.

Fixes #4626

westonruter commented on includes/amp-helper-functions.php:
```diff
@@ -318,6 +318,133 @@ function amp_add_frontend_actions() {
+	// @todo If redirected to this page because AMP is not available due to validation errors, we need to return false here!
+	if ( isset( $_GET['noamp'] ) && 'validation' === $_GET['noamp'] ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		return false;
+	}
```
We need to revisit this.

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -2092,42 +2078,19 @@ static function( Optimizer\Error $error ) {
-				return $response;
-			}
-		}
+			// @todo Is this the best? Can noamp be a standard way to indicate AMP not available? Put it in AMP_HTTP?
+			// Note the value of validation here has the result of preventing AMP from even being considered available.
+			$non_amp_url = add_query_arg( 'noamp', 'validation', $non_amp_url );
```
Needs to be revisited.

westonruter replied:
Note: This will have no effect in Standard mode, so it is fine to allow even for unauthenticated users.

westonruter commented on src/MobileRedirection.php:
```diff
@@ -0,0 +1,456 @@
+			// Redirect if mobile redirection is not disabled for the session and JS redirection is disabled.
+			if ( ! $this->redirection_disabled_for_session() && ! $this->should_redirect_via_js() ) {
+				if ( ! headers_sent() ) {
+					header( 'Vary: User-Agent' ); // @todo This needs to not replace existing.
```
Need to merge this with other existing `Vary` headers.

westonruter reviewed (commented):
This PR has evolved quite a bit and since the code is cutting across several concerns, it is ending up going to close several issues.

@schlessera Please take a look at how I've moved things into the `MobileRedirection` service and give your overall sense for how this PR is going.

schlessera commented on src/MobileRedirection.php:
```diff
@@ -0,0 +1,456 @@
+	 */
+	public function __construct() {
+		$this->disabled_cookie_is_set = false;
+	}
```
```suggestion
	private $disabled_cookie_is_set = false;
```

westonruter replied:
It looks like this was old comment as the code is no longer current. I removed the variable entirely.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/4479-mobile-redirection

* 'develop' of github.com:ampproject/amp-wp: (43 commits)
  Additional check in e2e beforEach function
  Nav e2e test fix
  Fix reader themes e2e test
  Re-require options context rest endpoint prop
  Break long import into separate lines
  Require technical background selection before moving forward
  Update comment explaining temporary code
  Lint fixes
  Remove unnecessary state initialization
  Remove unnecessary state initialization
  Update theme selection text when reader mode is not selected
  Merge changes from code review
  Remove unused AMPNotice icon prop
  Revert change to CSS file placement
  Fix PHPCS issues
  Add missing @return description
  Fix wrong namespace for tests
  Add tests for LazilyInstantiatedService
  Add tests for SimpleServiceContainer
  Add tests for ServiceBasedPlugin
  ...

westonruter committed:
Allow amp_mobile_user_agents to include regex patterns

westonruter committed:
Fix tests and improve should_redirect_via_js

westonruter committed:
Use sessionStorage for disabling client-side redirection

* Prevent redirection logic from being enabled if AMP-first.
* Refactor logic for handling redirection on non-AMP endpoint.
* Exit response when doing server-side redirection redirection.
* Pass entire AMP URL to JS rather than just the amp slug.
* Eliminate checking value of cookie or query param.
* Fix sending of Vary header when server-side redirects used.
* Improve method naming.
* Improve logic for setting and clearing cookies.

westonruter committed:
Ensure that AMP-to-AMP linking is enabled when mobile redirection is enabled

westonruter committed:
Move mobile version switcher CSS to separate CSS file

Print in head if not filtered off

westonruter committed:
Remove unused disabled_cookie_is_set class variable

westonruter committed:
Add rel=alternate link in addition to existing amphtml link on non-AMP version

westonruter committed:
Ensure mobile redirection is hidden initially if Standard mode is selected

westonruter commented on src/MobileRedirection.php:
```diff
@@ -0,0 +1,530 @@
+		 *
+		 * @param bool $should_redirect_via_js Whether JS redirection should be used to take mobile visitors to the AMP version.
+		 */
+		return (bool) apply_filters( 'amp_mobile_redirect_via_js', true ); // @todo Consider `amp_mobile_client_side_redirection`.
```
Is `amp_mobile_client_side_redirection` a better name for this filter?

westonruter replied:
Done in d90f65af4c16bb8e7ed552d1d4cf6859daba33ed.

westonruter commented on src/MobileRedirection.php:
```diff
@@ -0,0 +1,530 @@
+	 */
+	const DISABLED_COOKIE_NAME = 'amp_mobile_redirect_disabled';
+
+	/**
+	 * Query parameter to indicate that the page in question should not be served as AMP.
+	 *
+	 * @todo This needs to not be specific to mobile.
+	 *
+	 * @var string
+	 */
+	const NO_AMP_QUERY_VAR = 'noamp';
```
Given that `AMP_Validation_Manager` also needs to redirect to the non-AMP version with a `noamp` query param, perhaps this should not be located in this service.

schlessera replied:
I'd vote for a `QueryVar` interface (or class if logic is needed) with this and other constants.

schlessera replied:
Alternatively, we could create a `QueryVars` object that can be used as a dependency for any service that interacts with them.

This would let us abstract the entire logic of retrieving them, instead of only their name.

So, instead of this:
```php
$url = add_query_arg( amp_get_slug(), '1', amp_get_current_url() );
$url = remove_query_arg( self::NO_AMP_QUERY_VAR, $url );
return $url;
```

We would have something like this, for example:
```php
return $this->query_vars
	->add( QueryVars::SLUG, 1 )
	->remove( QueryVars::NO_AMP )
	->get_url();
```

The internal logic can then include and validation or safeguards, or compact logic.

westonruter commented on src/MobileRedirection.php:
```diff
@@ -0,0 +1,530 @@
+			} elseif ( ! $this->is_redirection_disabled_via_cookie() ) {
+				if ( $this->is_redirection_disabled_via_query_param() ) {
+					// Persist disabling mobile redirection for the session if redirection is disabled for the current request.
+					$this->set_cookie_to_disable_redirection();
```
When accessing the AMP version, should the reverse of `set_cookie_to_disable_redirection` be done when server-side redirection is being performed? Should the cookie get cleared? Currently it is getting cleared via the logic in the `onclick` attribute on the version switcher link.

westonruter replied:
So something like:

```php
if ( is_amp_endpoint() && ! $js && $this->is_redirection_disabled_via_cookie() ) {
    $this->remove_cookie_to_disable_redirection();
}

westonruter replied:
Which would go in the `else` statement below, naturally.

westonruter replied:
This would eliminate the need to disable `httponly` which may be advantageous.

westonruter replied:
Done in 7b46bae117a1c5432c972d4d0fa126f8bc4025af

westonruter commented on src/MobileRedirection.php:
```diff
@@ -0,0 +1,530 @@
+		$expires  = 0;                                             // Time till expiry. Setting it to `0` means the cookie will only last for the current browser session.
+		$path     = wp_parse_url( home_url( '/' ), PHP_URL_PATH ); // Path.
+		$secure   = is_ssl();                                      // Whether cookie should be transmitted over a secure HTTPS connection.
+		$httponly = false;                                         // Whether cookie should be made accessible only through the HTTP protocol.
```
This could be changed to `true` if we clear the cookie via PHP as mentioned above.

westonruter replied:
Done in 7b46bae117a1c5432c972d4d0fa126f8bc4025af

schlessera commented on assets/src/mobile-redirection.js:
```diff
@@ -0,0 +1,54 @@
+		return;
+	}
+
+	const regExpRegExp = new RegExp( regexRegex );
```
(⊙_⊙;)

westonruter replied:
I've added a comment above to explain the meta regex.

schlessera commented on includes/amp-helper-functions.php:
```diff
@@ -318,6 +318,133 @@ function amp_add_frontend_actions() {
+/**
+ * Determine whether AMP is available for the current URL.
+ *
+ * @todo This should go into the AmpProject\AmpWP namespace?
```
This depends on whether we want to have "templating tags" under namespaces.

I'm all for namespacing everything, but we have to keep in mind it will raise the learning curve for some people.

Alternatively, we can also think about putting functions like these within static classes, which should at least add autoloading as a benefit: `AmpWP::is_available()` or something like that.

johnwatkins0 replied:
@schlessera Curious about the learning curve -- do you find that many PHP developers in the WP space are not familiar with namespacing functions? It seems like I've been seeing it as a routine practice for a few years now wherever supporting old PHP is not required. I'd definitely consider it easier to pick up than dependency injection 😄

westonruter replied:
I'll just go with the global function for now, but I've opened #4946 to discuss further.

schlessera commented on includes/amp-helper-functions.php:
```diff
@@ -318,6 +318,133 @@ function amp_add_frontend_actions() {
+		'wp',
+		'WP_Query',
+		'amp_skip_post()'
+	);
```
I'd put this into a separate function and call the function from below in the multiple separate cases.

Right now, this will always incur a translation cost, even if it is not being used, which should actually be 99% of the cases.

westonruter replied:
Done in 367134d4f

schlessera commented on includes/amp-helper-functions.php:
```diff
@@ -318,6 +318,133 @@ function amp_add_frontend_actions() {
+	}
+
+	// Always return false when requesting the service worker.
+	// Note this is no longer required because AMP_Theme_Support::prepare_response() will abort for non-HTML responses.
```
Why not remove it right away, then?

westonruter replied:
It's not _required_ but it is still good to have because it avoids needlessly output buffering the response.

```suggestion
	// Note this is no longer strictly required because AMP_Theme_Support::prepare_response() will abort for non-HTML responses.
	// But it is still good to do so because it avoids needlessly output-buffering the response.
```

schlessera commented on includes/validation/class-amp-validation-manager.php:
```diff
@@ -1886,54 +1798,153 @@ public static function remove_illegal_source_stack_comments( Document $dom ) {
+			$dom->documentElement->removeAttribute( Attribute::AMP_EMOJI );
+			$dom->documentElement->removeAttribute( Attribute::AMP_EMOJI_ALT );
+
+			/*
+			 * Make sure that document.write() is disabled to prevent dynamically-added content (such as added
+			 * via amp-live-list) from wiping out the page by introducing any scripts that call this function.
+			 */
+			$script = $dom->createElement( Tag::SCRIPT );
+			$script->appendChild( $dom->createTextNode( 'document.addEventListener( "DOMContentLoaded", function() { document.write = function( text ) { throw new Error( "[AMP-WP] Prevented document.write() call with: "  + text ); }; } );' ) );
+			$dom->head->appendChild( $script );
+			return true;
```
This sounds like it might inadvertently break needed behavior. Are there any valid uses for `document.write` at this point?

westonruter replied:
Calling `document.write()` after the page has loaded is never valid. It causes the page to be destroyed. Example: https://jsfiddle.net/westonruter/hq98er6j/

This prevents a scenario where an `amp-live-list` is used on the page (in an AMP-first site), and a site has _kept_ an invalid script that does `document.write()`. When the new content for the `amp-live-list` is loaded, if there was a `document.write()` call it would break the page.

westonruter replied:
It's important to note that this function is only overridden at `DOMContentLoaded`, so it will still work properly when it should be.

schlessera commented on includes/validation/class-amp-validation-manager.php:
```diff
@@ -1886,54 +1798,153 @@ public static function remove_illegal_source_stack_comments( Document $dom ) {
+		$parent_menu_link = $dom->xpath->query( './a[ @href ]', $parent_menu_item )->item( 0 );
+		$admin_bar_icon   = $dom->xpath->query( './span[ @id = "amp-admin-bar-item-status-icon" ]', $parent_menu_link )->item( 0 );
+		$validate_link    = $dom->xpath->query( './/li[ @id = "wp-admin-bar-amp-validity" ]/a[ @href ]', $parent_menu_item )->item( 0 );
+		if ( ! $validate_link instanceof DOMElement || ! $admin_bar_icon instanceof DOMElement || ! $validate_link instanceof DOMElement ) {
```
```suggestion
		if ( ! $parent_menu_link instanceof DOMElement || ! $admin_bar_icon instanceof DOMElement || ! $validate_link instanceof DOMElement ) {
```

schlessera reviewed (changes requested):
General approach seems sound to me.

There are still a few todos, of course, and given some of them sound so innocent but haven't been just executed upon nevertheless makes me think there's still a bit of hidden complexity to uncover... hopefully, the PR doesn't become too unwieldy.

johnwatkins0 commented on includes/amp-helper-functions.php:
```diff
@@ -318,6 +318,133 @@ function amp_add_frontend_actions() {
+	$warned        = false;
+	$error_message = sprintf(
+		/* translators: %1$s: is_amp_endpoint(), %2$s: the current action, %3$s: the wp action, %4$s: the WP_Query class, %5$s: the amp_skip_post() function */
+		__( '%1$s was called too early and so it will not work properly. WordPress is currently doing the "%2$s" action. Calling this function before the "%3$s" action means it will not have access to %4$s and the queried object to determine if it is an AMP response, thus neither the "%5$s" filter nor the AMP enabled toggle will be considered.', 'amp' ),
```
Seeing this message on the login screen.

<img width="490" alt="Screen Shot 2020-06-25 at 10 47 02 AM" src="https://user-images.githubusercontent.com/9020968/85752783-50e7a580-b6d1-11ea-86a5-983565e78c50.png">

westonruter replied:
Fixed in e1d740f3bae9ce490c53435e09a03bae64eab763

westonruter committed:
Restrict honoring noamp=validation to when not in Standard mode

westonruter committed:
Rename amp_mobile_redirect_via_js filter

westonruter committed:
Fix duplicate/missing variable reference

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

westonruter committed:
Use httpOnly for server-side redirection disabled cookie

* Eliminate use of onclick attribute on link, putting JS-based redirection in mobile-redirection.js
* Use COOKIE_DOMAIN constant for cookie domain (as opposed to empty value).

westonruter committed:
Convert error_message into function

westonruter committed:
Explain why still good to short-circuit for SW requests

westonruter committed:
Fix unexpected _doing_it_wrong() when is_amp_endpoint() used on wp-login.php

* Limit _doing_it_wrong() to only emit from is_amp_available().
* Ensure is_amp_available() is always called in is_amp_endpoint().
* Fix _doing_it_wrong() message due to use of closure.

westonruter committed:
Remove todo now captured in #4946

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/4479-mobile-redirection

* 'develop' of github.com:ampproject/amp-wp: (27 commits)
  Fix issue of options and dev tools option being saved twice
  Update snapshot
  lint fixes
  Change phone ratio
  Summary screen ui updates
  Fix PHPCS issues
  Use SVG variant of `yes-alt` Dashicon
  Minor updates to summary screen
  Add dimensions to images and set min heights on containers
  Darken unselected radio input
  ESLint fix
  Fix a comment
  E2E test fixes
  Style updates
  phpstan fix
  Update an e2e test
  Make setup wizard links open in new tab
  Setup wizard: ignore defaults on first run
  Disable filtering of services by default
  Add additional tests for ServiceBasedPlugin
  ...

westonruter committed:
Make is_amp_available() return true before 'wp' action if Standard mode

* Add tests.
* Remove duplicate logic in is_amp_available()

westonruter committed:
Add QueryVars interface for capture amp/noamp and values

Let is_amp_available()/is_amp_endpoint() return true before wp action if theme supports AMP _and_ all templates are supported.

westonruter committed:
Remove noamp query var from URLs to AMP pages

westonruter committed:
Re-use is_amp_available() when adding items to admin bar

westonruter committed:
Prevent initializing redirect if functionality disabled

* Ensure that Vary header is sent when server-side redirection is used.
* Short-circuit logic when server-side redirection enabled and it's not an AMP page nor a mobile user agent.

westonruter committed:
Add missing QueryVars.php

westonruter committed:
Prevent mobile redirection in Customizer to avoid complications

westonruter committed:
Show mobile version switcher in Customizer, but disable clicks there and in Paired Browsing

westonruter committed:
Strip noamp query var when navigating in paired browsing

westonruter commented:
After banging on this for awhile, I believe the only thing left is tests for the `MobileRedirection` class.

westonruter committed:
Allow AMP to be disabled for a request via noamp=available when user can validate

Also add a "View with AMP disabled" link to the admin bar for such users to easily disable AMP for a URL when a site is in Standard mode without requiring them to switch the entire site to Transitional mode or to disable AMP for a given URL.

See #1294

westonruter committed:
Add full test coverage for MovileRedirection service

westonruter committed:
Simplify methods for printing mobile version switcher

westonruter committed:
Add amp_mobile_version_switcher_link_text filter
